src/memory/AmdFlash.cc

```diff
@@ -648,49 +670,115 @@ bool AmdFlash::checkCommandContinuityCheck()
 	return false;
 }
 
-bool AmdFlash::checkCommandEraseSector()
+bool AmdFlash::checkCommandEraseSector(EmuTime::param time)
 {
 	static constexpr std::array<uint8_t, 5> cmdSeq = {0xaa, 0x55, 0x80, 0xaa, 0x55};
 	if (partialMatch(cmdSeq)) {
 		if (cmd.size() < 6) return true;
 		if (cmd[5].value == 0x30) {
 			clearStatus();
 
-			if (const Sector& sector = getSector(cmd[5].addr);
+			if (Sector& sector = getSector(cmd[5].addr);
 			    isWritable(sector)) {
-				ram->memset(sector.writeAddress, 0xff, sector.size);
+				sector.erasePending = true;
 			} else {
 				status.sectorLock = true;
 			}
 
-			status.dataPolling = true; // immediate completion
+			status.ready = false;
+			status.dataPolling = false;
+			setState(State::ERASE_SECTOR);
+			scheduleEraseAlgorithm(time);
 		}
 	}
 	return false;
 }
 
-bool AmdFlash::checkCommandEraseChip()
+bool AmdFlash::checkCommandEraseAdditionalSector(EmuTime::param time)
+{
+	assert(state == State::ERASE_SECTOR);
+	if (!status.eraseTimer && cmd[0].value == 0x30) {
+		Sector& sector = getSector(cmd[0].addr);
+		if (isWritable(sector)) {
+			sector.erasePending = true;
+		} else {
+			status.sectorLock = true;
+		}
+
+		scheduleEraseAlgorithm(time);
+	}
+	return false;
+}
+
+bool AmdFlash::checkCommandEraseChip(EmuTime::param time)
 {
 	static constexpr std::array<uint8_t, 5> cmdSeq = {0xaa, 0x55, 0x80, 0xaa, 0x55};
 	if (partialMatch(cmdSeq)) {
 		if (cmd.size() < 6) return true;
 		if (cmd[5].value == 0x10) {
 			clearStatus();
 
-			for (const Sector& sector : sectors) {
+			for (Sector& sector : sectors) {
 				if (isWritable(sector)) {
-					ram->memset(sector.writeAddress, 0xff, sector.size);
+					sector.erasePending = true;
 				} else {
 					status.sectorLock = true;
 				}
 			}
 
-			status.dataPolling = true; // immediate completion
+			status.ready = false;
+			status.dataPolling = false;
+			setState(State::ERASE_CHIP);
+			scheduleEraseAlgorithm(time);
 		}
 	}
 	return false;
 }
 
+void AmdFlash::scheduleEraseAlgorithm(EmuTime::param time)
+{
+	if (state == State::ERASE_SECTOR && !status.eraseTimer) {
+		syncAlgorithm.removeSyncPoint();
+		syncAlgorithm.setSyncPoint(time + chip.erase.timerDuration);
+	} else if (std::ranges::any_of(sectors, [](auto& s) { return s.erasePending.value_or(false); })) {
```

Comment:
Very much a detail: I'd expected `const auto&` parameter here.

And in some other place (not in this patch) I now see that you pass `auto` (instead of `const auto&`).

Not that this will many any difference in the generated code, but it's good to express the intention to human readers of the code.